pkg/webhook/vm_validatation.go

```diff
@@ -108,3 +123,67 @@ func (vmValidator *vmDeviceHostValidator) validateDevicesFromSameNodes(vmObj *ku
 
 	return nil
 }
+
+func (vmValidator *vmDeviceHostValidator) validateHostDevices(vmObj *kubevirtv1.VirtualMachine) error {
+	for _, hostDevice := range vmObj.Spec.Template.Spec.Domain.Devices.HostDevices {
+		var (
+			foundInPCI, foundInUSB bool
+			err                    error
+		)
+
+		if foundInPCI, err = vmValidator.validatePCIDevice(hostDevice.DeviceName); err != nil {
+			return err
+		}
+
+		if foundInUSB, err = vmValidator.validateUSBDevice(hostDevice.DeviceName); err != nil {
+			return err
+		}
+
+		if !foundInPCI && !foundInUSB {
+			return fmt.Errorf("hostdevice %s: resource name %s not found in pcidevice and usbdevice cache", hostDevice.Name, hostDevice.DeviceName)
+		}
+	}
+	return nil
+}
+
+func (vmValidator *vmDeviceHostValidator) validatePCIDevice(resourceName string) (found bool, err error) {
+	pciDeviceObjs, err := vmValidator.pciCache.GetByIndex(PCIDeviceByResourceName, resourceName)
+	if err != nil {
+		return false, fmt.Errorf("error looking up pcidevice %s from cache: %v", resourceName, err)
+	}
+
+	if len(pciDeviceObjs) == 0 {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (vmValidator *vmDeviceHostValidator) validateUSBDevice(resourceName string) (found bool, err error) {
+	usbDeviceObjs, err := vmValidator.usbCache.GetByIndex(USBDeviceByResourceName, resourceName)
+	if err != nil {
+		return false, fmt.Errorf("error looking up usbdevice %s from cache: %v", resourceName, err)
+	}
+
+	if len(usbDeviceObjs) == 0 {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (vmValidator *vmDeviceHostValidator) validateGPUs(vmObj *kubevirtv1.VirtualMachine) error {
+	for _, gpu := range vmObj.Spec.Template.Spec.Domain.Devices.GPUs {
+		foundInPCI, err := vmValidator.validatePCIDevice(gpu.DeviceName)
+
+		if err != nil {
+			return err
+		}
+
+		if !foundInPCI {
+			return fmt.Errorf("gpu device %s: resource name %s not found in pcidevice cache", gpu.Name, gpu.DeviceName)
+		}
```

Comment:
Hi Jack,
Why do we validate GPUs with PCI devices?

Reply:
As I know, the GPU is also bound to PCI device, so we can use PCI device to check it.

Reply:
I thought we just mapped the vGPU to PCI device. It's like USB device. We can pass whole USB controller to VM with PCI device, or just pass one of USB ports to VM with USB passthrough.